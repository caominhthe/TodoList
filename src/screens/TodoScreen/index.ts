export * from "./TodoScreen";
export * from "./store";
