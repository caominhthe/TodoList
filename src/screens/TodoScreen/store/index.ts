export * from "./todo.interface";
export * from "./todo.slice";
export * from "./todo.selector";
