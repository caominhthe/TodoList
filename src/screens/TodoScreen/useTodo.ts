import { useAppDispatch } from "@root/src/hooks/useAppDispatch";
import {
  Todo,
  createTodo as createAction,
  deleteTodo as deleteAction,
  updateTodo,
} from "./store";
import { useCallback } from "react";

export const useToDo = () => {
  const dispatch = useAppDispatch();

  //Clear input content Items after create new todo item
  const createTodo = useCallback(
    (content: string) => {
      if (content !== "") {
        dispatch(createAction(content.trim() as string));
      }
    },
    [dispatch]
  );

  //Clear input content and selected Items after Update new todo item
  const updateToDo = useCallback(
    (content: string, selectedTodo: Todo) => {
      if (selectedTodo && content !== "") {
        const updatedTodo: Todo = {
          ...selectedTodo,
          title: content.trim(),
        };
        dispatch(updateTodo(updatedTodo));
      }
    },
    [dispatch]
  );

  // Reset content input and inputMode after delete to avoid confuse when edit
  const deleteTodo = useCallback((todo: Todo) => {
    dispatch(deleteAction(todo.id));
  }, []);

  return { createTodo, updateToDo, deleteTodo };
};
