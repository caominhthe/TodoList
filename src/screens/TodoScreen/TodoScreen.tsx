import React, { useCallback, useRef, useState } from "react";
import {
  FlatList,
  StyleSheet,
  View,
  Text,
  TextInput,
  Keyboard,
} from "react-native";
import { useSelector } from "react-redux";
import { RootStackScreenProps, ScreenNames } from "@routers";
import { KeyboardAvoidContainer, TodoInputGroup, TodoItem } from "@components";
import { EditMode, Todo, selectTodoEditMode, selectTodos } from "./store";
import { Palette, spacing, typography } from "@theme";
import { useToDo } from "./useTodo";
import { Colors } from "@root/src/theme/color";

type TodoScreenProps = RootStackScreenProps<ScreenNames.Todo>;

export const TodoScreen: React.FC<TodoScreenProps> = () => {
  const [selectedTodo, setSelectedTodo] = useState<Todo | null>();
  const [content, setContent] = useState("");
  const inputMode = useSelector(selectTodoEditMode);
  const todoList: Todo[] = useSelector(selectTodos);
  const isAddMode = inputMode === EditMode.Add;
  const inputRef = useRef<TextInput>(null);

  const { createTodo, updateToDo, deleteTodo, changeInputMode } = useToDo();

  //Clear input content Items after create new todo item
  const onCreateTodo = useCallback(() => {
    if (content !== "") {
      createTodo(content);
      clearInput();
    }
  }, [content]);

  //Clear input content and selected Items after Update new todo item
  const onUpdateTodo = useCallback(() => {
    if (selectedTodo && content !== "") {
      updateToDo({ title: content.trim() }, selectedTodo.id);
      clearInput();
    }
  }, [selectedTodo, content]);

  // Reset content input and inputMode after delete to avoid confuse when edit
  const onPressDelete = useCallback(
    (todo: Todo) => {
      deleteTodo(todo);
    },
    [deleteTodo]
  );

  // Input turn into edit inputMode when click on an item, selected item will be highlight
  const onPressItem = useCallback(
    (todo: Todo) => {
      setSelectedTodo(todo);
      setContent(todo.title);
      changeInputMode(EditMode.Edit);
      inputRef.current?.focus();
    },
    [inputMode]
  );

  const onPressCheckBox = useCallback((todo: Todo) => {
    updateToDo({ isDone: !todo.isDone }, todo.id);
  }, []);

  const clearInput = () => {
    setSelectedTodo(null);
    setContent("");
    changeInputMode(EditMode.Add);
    Keyboard.dismiss();
  };

  //If cancel keyboard during edit, it will cancel the edit mode
  const cancelEdit = () => {
    if (!isAddMode) {
      clearInput();
    }
  };

  const renderItem = (todo: Todo) => {
    return (
      <TodoItem
        todo={todo}
        onPress={onPressItem}
        onDelete={onPressDelete}
        onPressCheckBox={onPressCheckBox}
        isSelected={todo.id === selectedTodo?.id}
      />
    );
  };

  return (
    <KeyboardAvoidContainer withSafeKeyboard>
      <FlatList<Todo>
        showsVerticalScrollIndicator={false}
        data={todoList}
        keyExtractor={(todo) => String(todo.id)}
        renderItem={({ item }) => renderItem(item)}
        contentContainerStyle={styles.flatListContent}
      />
      <View style={styles.inputContainer}>
        <TodoInputGroup
          value={content}
          onChangeText={(text) => setContent(text)}
          onPressButton={isAddMode ? onCreateTodo : onUpdateTodo}
          onBlur={cancelEdit}
          inputRef={inputRef}
          button={
            <Text style={styles.inputButton}>{isAddMode ? "Add" : "Edit"}</Text>
          }
        />
      </View>
    </KeyboardAvoidContainer>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    backgroundColor: Palette.white,
  },
  flatListContent: {
    flexGrow: 1,
    paddingBottom: spacing.large,
    paddingTop: spacing.medium,
  },
  keyboardAvoidView: {
    flex: 1,
    justifyContent: "space-between",
  },
  inputContainer: {
    borderTopWidth: 1,
    borderColor: Palette.grey,
    paddingTop: spacing.small,
  },
  inputButton: {
    fontWeight: "bold",
    fontSize: typography.fontSize.medium,
    backgroundColor: Colors.Background,
    padding: spacing.medium,
    borderRadius: spacing.small,
    overflow: "hidden",
  },
});
