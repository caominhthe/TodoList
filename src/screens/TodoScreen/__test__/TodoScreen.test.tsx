import mockSafeAreaContext from "react-native-safe-area-context/jest/mock";
import { render } from "@testing-library/react-native";
import * as ReactRedux from "react-redux";
import { TodoScreen } from "../TodoScreen";
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);
const mockNavigate = jest.fn();
const baseProps = {
  route: {},
  navigation: {
    navigate: mockNavigate,
  },
} as any;
jest.mock("react-native-safe-area-context", () => mockSafeAreaContext);

const mockTodo = {
  id: "id",
  title: "test",
  createdAt: "1690031534767",
  isDone: false,
};

describe("Given Todo Screen", () => {
  let store: any;

  it("should render correctly with todo items", () => {
    store = mockStore({
      todo: {
        todoList: [mockTodo],
      },
    });
    const { toJSON } = render(
      <ReactRedux.Provider store={store as any}>
        <TodoScreen {...baseProps} />
      </ReactRedux.Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
