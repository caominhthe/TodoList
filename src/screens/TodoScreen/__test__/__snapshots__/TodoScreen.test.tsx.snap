// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given Todo Screen should render correctly with todo items 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
          "justifyContent": "space-between",
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "paddingBottom": 24,
          "paddingTop": 12,
        }
      }
      data={
        [
          {
            "createdAt": 1689951787844,
            "id": "id",
            "title": "test",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "borderColor": "grey",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "marginHorizontal": 24,
                  "marginTop": 8,
                  "minHeight": 40,
                  "paddingHorizontal": 24,
                  "paddingTop": 8,
                  "paddingVertical": 8,
                },
                false,
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "flex": 1,
                  "opacity": 1,
                }
              }
              testID="todo-item"
            >
              <Text
                style={
                  {
                    "flexWrap": "wrap",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "paddingHorizontal": 12,
                  }
                }
                testID="item-title"
              >
                test
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="delete-button"
            >
              <Text
                style={
                  {
                    "color": "blue",
                    "fontSize": 16,
                  }
                }
              >
                Delete
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        {
          "borderColor": "#DADADA",
          "borderTopWidth": 1,
          "paddingTop": 8,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "paddingBottom": 4,
            "paddingHorizontal": 12,
            "width": 750,
          }
        }
      >
        <TextInput
          multiline={true}
          onChangeText={[Function]}
          placeholder="Add new todo"
          style={
            {
              "borderColor": "black",
              "borderWidth": 1,
              "flex": 1,
              "flexShrink": 1,
              "fontSize": 20,
              "fontWeight": "700",
            }
          }
          value=""
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="submit-button"
        >
          <Text
            style={
              {
                "backgroundColor": "#e5e5e5",
                "borderRadius": 8,
                "fontSize": 16,
                "fontWeight": "bold",
                "overflow": "hidden",
                "padding": 12,
              }
            }
          >
            Edit
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
